designs/statefulset-disruption.md

```diff
@@ -0,0 +1,739 @@
+
+If Karpenter calls EC2 TerminateInstance **before** EC2 DetachVolume calls finish, then the volumes won't be detached **until the old instance terminates**. This delay depends on the instance type. 1 minute for `m5a.large`, up to 15 minutes for certain GPU/Windows instances (especially painful for ML/AI customers).
+
+Operators will see `FailedAttachVolume` events with `Multi-Attach error` and then `AttachVolume.Attach failed` errors. 
+
+### Solution B1: Wait for unmounts in Karpenter cloudProvider.Delete
+
+Wait for volumes to unmount before terminating the instance. 
+
+We can do this by waiting for all volumes of drain-able nodes to be marked as not be in use before terminating the node in c.cloudProvider.Delete (until a maximum of 20 seconds). 
+
+This means that our sequence of events will match the ideal diagram from section [### Ideal Graceful Shutdown for Stateful Workloads]
+
+We can use similar logic to [today's proof-of-concept implementation](https://github.com/kubernetes-sigs/karpenter/pull/1294), but move it to karpenter-provider-aws and check for `node.Status.VolumesInUse` instead of listing volumeattachment objects. A 20 second max wait was sufficient to prevent delays with m5a instance type, but further testing is needed to ensure it is enough for Windows/GPU instance types.
+
+**Pros:**
+- Leaves decision to each cloud/storage provider 
+- Can opt-in to this behavior for specific CSI Drivers (Perhaps via Helm parameter)
+- Only delays termination of nodes with stateful workloads.
+**Cons:**
+- Delays node termination and finalizer deletion by a worst-case of 20 seconds. (We can skip waiting on the volumes of non-drainable pods to make the average case lower)
+- Other CSI Drivers must opt-in
+
+### Alternatives Considered
+
+**Implement B1 in `kubernetes-sigs/karpenter`**
+
+Instead of solving this inside the [`c.cloudProvider.Delete`](https://github.com/aws/karpenter-provider-aws/blob/9cef47b6df77ec9e0a39dc6f4a4ecd1aab504ae3/pkg/cloudprovider/cloudprovider.go#L179), solve this inside [termination controller's reconciler loop](https://github.com/kubernetes-sigs/karpenter/blob/94d5b41d4711b1c21fc0264f1b29db6a64b95caf/pkg/controllers/node/termination/controller.go#L77) (as is done in [today's proof-of-concept implementation](https://github.com/kubernetes-sigs/karpenter/pull/1294))
+
+**Pros:**
+- Karpenter-provider-aws does not need to know about Kubernetes volume lifecycle
+**Cons:**
+- [Open Question] EBS may be the only storage provider where detaching volumes before terminating the instance matters. If this is the case, the delay before instance termination is not worth it for customers of other storage/cloud providers.
+- Should not hardcode cloud-provider specific CSI Drivers in upstream project. 
+
+**Karpenter Polls EC2 for volume detachments before calling EC2 TerminateInstance**
+
+Karpenter-provider-aws can poll EC2 DescribeVolumes before making an EC2 TerminateInstance call. 
+
+**Pros:**
+- Solves issue
+**Cons:**
```

Comment:
needs and extra line or double space at the end of the previous line, to properly break line in MD